timesketch/lib/analyzers/geoip.py

```diff
@@ -0,0 +1,311 @@
+    """Sketch analyzer for geolocating IP addresses.
+
+    Concrete plugin implementations should define the following attributes
+    NAME, DISPLAY_NAME, DESCRIPTION, GEOIP_CLIENT.
+    """
+
+    NAME = ""
+    DISPLAY_NAME = ""
+    DESCRIPTION = ""
+
+    GEOIP_CLIENT = None
+
+    DEPENDENCIES = frozenset(['feature_extraction'])
+    IP_FIELDS = ['ip', 'host_ip', 'src_ip', 'dst_ip', 'source_ip', 'dest_ip',
+        'ip_address', 'client_ip', 'address', 'saddr', 'daddr',
+        'requestMetadata_callerIp', 'a_answer']
+
+
+    def __init__(self, index_name, sketch_id, timeline_id=None):
+        """Initialize The Sketch Analyzer.
+        Args:
+            index_name: Elasticsearch index name
+            sketch_id: Sketch ID
+        """
+        self.index_name = index_name
+        super().__init__(index_name, sketch_id, timeline_id=timeline_id)
+
+
+    def _validate_ip(self, ip_address):
+        """Return true if ip_address is in a valid format and is non-private."""
+        try:
+            ip = ipaddress.ip_address(ip_address)
+            return ip.is_global
+        except ValueError as exception:
+            return False
+
+    def run(self):
+        """Entry point for the analyzer.
+
+        Returns:
+            String with summary of the analyzer result
+        """
+        query = ' OR '.join(
+            ['{0}:*'.format(ip_field) for ip_field in self.IP_FIELDS])
+
+        return_fields = self.IP_FIELDS.copy()
+
+        events = self.event_stream(
+            query_string=query,
+            return_fields=return_fields
+        )
+
+        try:
+            with self.GEOIP_CLIENT() as client:
+                ip_addresses = defaultdict(lambda: defaultdict(list))
+
+                for event in events:
+                    for ip_address_field in self.IP_FIELDS:
+                        ip_address = event.source.get(ip_address_field)
+                        if ip_address is None:
+                            continue
+
+                        ip_address = [ip_address] if isinstance(ip_address, str) else ip_address
+
+                        for ip_addr in ip_address:
+                            if not self._validate_ip(ip_addr):
+                                logger.debug('Value {0} in {1} not valid.'
+                                    .format(ip_addr, ip_address_field))
+                                continue
+                            ip_addresses[ip_addr][ip_address_field].append(event)
+
+                logger.info('Found {0} ip address(es)'.format(len(ip_addresses)))
+
+                for ip_address, ip_address_fields in ip_addresses.items():
+                    response = client.ip2geo(ip_address)
+
+                    if response is None:
+                        continue
+
+                    if len(response) != 5:
+                        logging.error('Invalid response from GeoIP client')
+
+                    iso_code, latitude, longitude, country_name, city_name = response
```

Comment:
You could even make the error message clearer: `Error: GegoIP client must return 5 fields: <iso_code, latitude, longitude, country_name, city_name>. Number of fields returned: X`

I would do something like:
```python
try:
  iso_code, latitude, longitude, country_name, city_name = response
except ValueError:
  logger.error('Error: GegoIP client must return 5 fields: <iso_code, latitude, longitude, country_name, city_name>. Number of fields returned: {0:d}'.format(len(response))
```

Reply:
Agree it could be clearer.  Implemented suggestion.